feat(blog) Navigating Access Control Design: Pursuing Clarity and Simplicity (#1883)
questdb/questdb.io: blog/2023-08-22-navigating-access-control-design-clarity-simplicity.mdx

description: Many consider access control lists a solved problem. But there still room for innovation. Read the article to found out where.
tags: [engineering, questdb, enterprise, rbac, acl]
image: /img/blog/2023-08-21/interitance-hell.webp
slug: navigating-access-control-design-clarity-simplicity
---

import Banner from "@theme/Banner"
import Screenshot from "@theme/Screenshot"

```
<Banner
  alt="tbd"
  height={591}
  src="/img/blog/2023-07-06/banner.webp"
  width={770}
/>
```

> QuestDB is a time-series database that offers fast ingest speeds, ILP and PGWire support and SQL query syntax. We recently shipped role-based access control (RBAC) in [QuestDB Enterprise](/enterprise/), and wanted to share the lessons we learned. Enjoy!

## Innovating Access Control

When delving into access control system design, one might assume that it is a solved problem and that all the important decisions have already been made. After all, many successful applications already exist, and each of them has solved its own access control challenge. While there is no need to re-invent the wheel,  there are many places  where innovation can still occur.

> In simple terms, when a user authenticates themselves, the system reads their permissions, then determines the operations that they are allowed to perform. To help with this, users are often categorized into groups or assigned specific roles, and thus synchronize their permissions with a wider bucket.

Author: ```suggestion
Let's start with an ACL in simple terms. When a user authenticates themselves, the system reads their permissions and then determines the operations that they are allowed to perform. To help with this, users are often categorized into groups or assigned specific roles, and thus synchronize their permissions with a wider bucket.
```

Our first dilemma lies in naming the entity that we use to organize users into collectives. Role seems like a good choice. A role implies that functionality is bundled and then granted to relevant users who fit the role's description. Roles are created, assigned to users, and their tasks are defined. In a perfect world, it's neat and tidy.

<Screenshot
  alt=""
  title="Roles, in Utopia"
  src="/img/blog/2023-08-21/roles-utopia.webp"
/>

However, here's the catch — unfortunately, reality falls short of perfection!

The initial simplicity soon crumbles as applications evolve, giving rise to a tangled and overlapping assortment of roles. These roles may then stray from their original purpose and lose their authentic alignment with their intended functionalities. Instead, they morph into a jumble of permissions for diverse users.

An analyst might be part developer. One developer may need a single permission that would be destructive in the hands of other another developer. And then you need another admin, but they're only half an admin. It starts clean, but complexity soon catches up. The challenge lies in maintaining clean roles amid this complexity.

<Screenshot

>   alt=""

Author: ```suggestion
  alt="Already things are wild. Only 5 uses, but 6 roles, a few are shared, some are special. Reality is much messier than Utopia."
```

  title="Roles, in reality"
  src="/img/blog/2023-08-21/roles-reality.webp"
/>

While roles  appear attractive at first, the term groups tends to offer more flexibility and thus, in theory, greater clarity. Consequently, the integration of both groups and roles often leads to even bigger confusion when inheritance is involved.

## Exploring Inheritance

Undoubtedly, groups are essential for efficient access control. Organized groups simplify the process of permission distribution. Access can be granted to a group of users in a single statement. Adding users into that group to then inherit those permissions makes intuitive sense. 

However, the necessity of nested group hierarchies warrants a second look. The question is whether to construct an intricate group hierarchy or be content with a singular level of inheritance. 

The risk of an intricate hierarchy is that it has the potential to escalate into an unmanageable tree-like structure, or even a complex graph,  if the system allows it. These configurations will soon perplex even the most skilled administrators.

Just take a quick look at the example below:

<Screenshot

>   alt=""

Author: ```suggestion
  alt="Groups, roles, 6 users now. But...8 groups!? Multiple permissions per group. Overlapping permissions. Ack! Help!"
```

## Considering Service Accounts

Beyond an administrator's organizational preferences, the application using the database will present its own critical questions that one must consider. For example, should a clear distinction exist between individual and application accounts?

The database logic alone might not necessitate any distinction.  But the application accounts may benefit from strict limitations to prevent inadvertent misuse.  A common trap is when an application inherits an overly-permissive capability by accident. Depending on how your groups or roles are arranged, it can be easy to do. We need something to help ensure that we do not make this mistake

The creation of service accounts serves this purpose.  Service accounts mirror standard users, differing only in their exclusion from group affiliations. Their access is explicitly defined, which prevents unintentional overreach and mis-assignment.

Here is a sample arrangement that shows groups, users, and service accounts:

<Screenshot

>   alt=""

Author: ```suggestion
  alt="Demonstrates service accounts. Separating human usage from system usage. It's busy, but still organized and rational. Nothing like inheritance hell! "
```

  title="Service Accounts"
  src="/img/blog/2023-08-21/service-accounts.webp"
/>

And the diagram below illustrates how the above users and service accounts could be used to establish connections with the database:

<Screenshot

>   alt=""

Author: ```suggestion
  alt="Showing groups and service accounts accessing a DB. Some people use PG Wire, some a REST API, and others ILP. Their permissions are organized and it's refreshing. Also nothing like inheritance hell."
```

---
title: "Navigating Access Control Design: Pursuing Clarity and Simplicity"
author: Imre Aranyosi
author_title: QuestDB Team
author_url: https://github.com/glasstiger
author_image_url: https://avatars.githubusercontent.com/glasstiger
description:
  Many consider access control lists a solved problem. But there still room for
  innovation. Read the article to found out where.
tags: [engineering, questdb, enterprise, rbac, acl]
image: /img/blog/2023-08-21/life-of-an-admin.webp
slug: navigating-access-control-design-clarity-simplicity
buttonText: "See the ACL on QuestDB Cloud"
---

import Banner from "@theme/Banner"
import Screenshot from "@theme/Screenshot"

> <Banner alt="tbd" src="/img/blog/2023-08-21/life-of-an-admin.webp" />

Reviewer: ```suggestion
<Banner alt="LEGO minifigures in a wide variety of costumes waiting in line in front of an office desk." src="/img/blog/2023-08-21/life-of-an-admin.webp" />
```
